common/src/escrow_accounts.rs

```diff
@@ -13,18 +16,84 @@ use tracing::{error, warn};
 
 use crate::prelude::{Query, SubgraphClient};
 
+#[derive(Clone, Debug, Default, PartialEq, Eq)]
+pub struct EscrowAccounts {
+    senders_balances: HashMap<Address, U256>,
+    signers_to_senders: HashMap<Address, Address>,
+    senders_to_signers: HashMap<Address, Vec<Address>>,
+}
+
+impl EscrowAccounts {
+    pub fn new(
+        senders_balances: HashMap<Address, U256>,
+        senders_to_signers: HashMap<Address, Vec<Address>>,
+    ) -> Self {
+        let signers_to_senders = senders_to_signers
+            .iter()
+            .flat_map(|(sender, signers)| signers.iter().map(move |signer| (*signer, *sender)))
+            .collect();
+
+        Self {
+            senders_balances,
+            signers_to_senders,
+            senders_to_signers,
+        }
+    }
+
+    pub fn get_signers_for_sender(&self, sender: &Address) -> Result<Vec<Address>> {
+        self.senders_to_signers
+            .get(sender)
+            .filter(|signers| !signers.is_empty())
+            .ok_or(anyhow::format_err!(
+                "No signers found for sender {}.",
+                sender
+            ))
+            .map(|signers| signers.to_owned())
+    }
+
+    pub fn get_sender_for_signer(&self, signer: &Address) -> Result<Address> {
+        self.signers_to_senders
+            .get(signer)
+            .ok_or(anyhow::format_err!(
+                "Sender not found for receipt signer {}.",
+                signer
+            ))
+            .copied()
+    }
+
+    pub fn get_balance_for_sender(&self, sender: &Address) -> Result<U256> {
+        self.senders_balances
+            .get(sender)
+            .ok_or(anyhow::format_err!(
+                "Balance not found for sender {}.",
+                sender
+            ))
+            .copied()
+    }
+
+    pub fn get_balance_for_signer(&self, signer: &Address) -> Result<U256> {
+        self.get_sender_for_signer(signer)
+            .and_then(|sender| self.get_balance_for_sender(&sender))
+            .map_err(|e| anyhow::format_err!("Could not get balance for signer {}: {}", signer, e))
+    }
```

Comment:
Can we return an `AdapterError` for all of these?

Is there any reason to map to anyhow?

Reply:
Yes because it's not part of the adapters code, and is used outside of adapters in some places. Though it's really only a choice, nothing stops us for throwing `AdapterError`s anyway, what do you think?

Reply:
I'd rather use AdapterError here. All the code would be much cleaner, also, I'm not sure why `thiserror` is not being so used. 

This could be a different error like `EscrowAccountsError` and used by thiserror to convert to AdapterError automatically.

This would address your concern about not being an AdapterError while helping with ergonomics to remove the 'map_err()'